lib/charms/mongodb/v0/shards_interface.py

```diff
@@ -264,6 +337,80 @@ def _on_relation_changed(self, event):
 
         # TODO future PR, leader unit verifies shard was added to cluster (update-status hook)
 
+    def _on_relation_broken(self, event) -> None:
+        """Waits for the shard to be fully drained from the cluster."""
+        if self.charm.is_role(Config.Role.REPLICATION):
+            self.unit.status = BlockedStatus("role replication does not support sharding")
+            logger.error("sharding interface not supported with config role=replication")
+            return
+
+        if not self.charm.is_role(Config.Role.SHARD):
+            logger.info(
+                "skipping relation broken event ShardingProvider is only be executed by shards"
+            )
+            return
+
+        if not self.charm.db_initialised:
+            event.defer()
+            return
+
+        self.charm.unit.status = MaintenanceStatus("Draining shard from cluster")
+        # mongos hosts must be retrieved via relation data, as relation.units are not available in
+        # broken
+        mongos_hosts = json.loads(event.relation.data[event.relation.app].get(HOSTS_KEY))
+        drained = False
+        while not drained:
+            try:
+                # no need to continuously check and abuse resources while shard is draining
+                time.sleep(10)
```

Comment:
just curious why we sleep frist?

Reply:
Consider having sleep last:
```
       try:
               drained = self.drained(mongos_hosts, self.charm.app.name)
               draining_status = (
                    "Shard is still draining" if not drained else "Shard is fully drained."
                )
                logger.debug(draining_status)
                time.sleep(10)
      except ... : 
                log and retry
```
If the commands below it fail, the `sleep(10)` is skipped. Thus the check repetitively happens and fails without waiting in between. Ideally if there is a failure we should sleep in between failures so that we don't abuse system resources.

Reply:
Got it, thank you